src/Mono.Android/Microsoft.Android.Runtime/ManagedValueManager.cs

```diff
@@ -152,23 +179,27 @@ void WarnNotReplacing (int key, IJavaPeerable ignoreValue, IJavaPeerable keepVal
 
 	public override void RemovePeer (IJavaPeerable value)
 	{
-		if (RegisteredInstances == null)
-			throw new ObjectDisposedException (nameof (ManagedValueManager));
+		ThrowIfDisposed ();
+
+		// Remove any collected contexts before modifying RegisteredInstances
+		CollectPeers ();
 
 		if (value == null)
 			throw new ArgumentNullException (nameof (value));
 
-		int key = value.JniIdentityHashCode;
 		lock (RegisteredInstances) {
-			List<IJavaPeerable>? peers;
-			if (!RegisteredInstances.TryGetValue (key, out peers))
+			int key = value.JniIdentityHashCode;
+			if (!RegisteredInstances.TryGetValue (key, out List<ReferenceTrackingHandle>? peers))
 				return;
 
 			for (int i = peers.Count - 1; i >= 0; i--) {
-				var p   = peers [i];
-				if (object.ReferenceEquals (value, p)) {
+				var peer = peers [i];
+				var target = peer.Target;
```

Comment:
Could you explicitly type `peer` and `target`? Using `var` in this context hides information, somebody not familiar with the code won't know what the types are, without looking elsewhere in the code.